### ISSUE TITLE
👍🏻 Save time and improve DevX with a more streamlined PR flow

### ISSUE THREAD

**Issue by janhalen** [OS2fleetoptimiser/OS2fleetoptimiser#32]
I have a suggestion for a way to improve the way new code is merged into the main branch.

1. Let the main branch be to only protected branch that requires a thorough PR flow with reviews.
2. Assemble smaller additions or fixes in logical issues with task lists for each change and make a bit longer running branches to fix and test more issues.
3. Merge these larger branches into main with a proper and maybe more thorough PR process.

The goal would be to have fewer PRs, enable direct commits of more fixes/additions to the branches.

Right now it seems that the PRs contains few comments and are all more or less directly merged, so it seems like extra load to the process without any benefits?

@andreasDroid  Would such a flow lower the cognitive load and time spent for the developers or are there to few issue at this time for it to make an impact? If so we could reconsider later if the load increases.

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/about-task-lists

https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/adding-sub-issues

**Comment by andreasDroid** [comment]
I partially agree with the approach, however for bug fixes we will not be waiting for bigger chunks of work to be done before merging to main. We want to quickly be able to push fixes that implicates the user experience.

For planned (feature) work we can surely do this and collect more code in a branch before making PRs. Time spent for the reviewer is of concern, since it'll be a bigger hassle to review PRs that include changes to many files.

**Comment by andreasDroid** [comment]
Hi again, @janhalen 

We've had some time to discuss this internally, and come to the conclusion that we can handle this in a better way.
We have had good experience with the more or less defined flow [gitflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) .

This would require us to have a development branch, to where we can push finished features. 
We like this workflow better for multiple reasons:
- we can have the latest stable release going and still be able to merge out hotfix branches for the main branch 
- we can work on bigger releases from the development branch where each pr to development is essentially a _working_ version of the code
- review process for new features is moved from branch --> main to branch --> development, providing an easier job for the reviewer. (We prefer PRs that are small and very specific over large junks) 
- release branches will be a historic view of a specific release

The development branch should have review rules for merging/PRs. 
Let me hear what you think, and if you can see the potential in this approach.

**Comment by janhalen** [comment]
Hey @andreasDroid, looking into your suggestions first thing next week and i'll get back to you with some comments. 
Hope you have a nice weekend!

**Comment by andreasDroid** [comment]
Hi again, @janhalen we are currently constructing a specific guide for how we would like to control git. The link I shared is legacy, and I will share with you our thoughts later this week. It's mostly similar to trunk-based development.

**Comment by janhalen** [comment]
Ok.. I'll wait.

The most important part for ensuring transparency is that there are clearly defined and well described issues in the repository linked to the work being done. (Preferably made from issue templates to ensure uniformity)

Next would be using SemVer and uniform commit messages (preferably conventional commits)

After that using tags and releases for the resulting work.

**Comment by andreasDroid** [comment]
Do you have issue templates that you usually use?

**Comment by andreasDroid** [comment]
Can you please add _allandroidsagency_ as maintainer in the organisation?

**Comment by janhalen** [comment]
> Can you please add _allandroidsagency_ as maintainer in the organisation?

Just to understand better, is this for automation purposes?

The GitHub TOS does not allow anonymous account, but recommends using TOKENS for automation.

https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#3-account-requirements

**Comment by janhalen** [comment]
> > Can you please add _allandroidsagency_ as maintainer in the organisation?
> 
> Just to understand better, is this for automation purposes?
> 
> The GitHub TOS does not allow anonymous account, but recommends using TOKENS for automation.
> 
> https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#3-account-requirements

@andreasDroid, Ahh silly me, my thoughts were elsewhere in full "audit mode".. Sorry...

Thinking about this task, I have a susggestion.

How about, to make os2fleetoptimiser even more sustainable and independent, that @sobuos and I use this as an exercise next time she is in? I can show her how to manage the RBAC on the org and repo? @sobuos are you up for that?

**Comment by sobuos** [comment]
@janhalen I will be in this Wednesday, should we look into it then? I am not sure I totally understand what it is about, but hopefully you can explain it to me!

**Comment by janhalen** [comment]
@sobuos See you wednesday, Ill explain it all..

**Comment by janhalen** [comment]
Hi @andreasDroid & @sobuos 

Great workflow outline suggestion! Lets go with that!

In addition, to enhance transparency and traceability, I would recommend that we also describe in the CONTRIBUTING.md how work should be tracked in public issues by making sure that PRs are always linked to a related issue. This practice ensures traceability and helps maintain a clear record of changes. To make it easy and simple for the devs we can [utilize keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue) to avoid duplication of effort of having to describe the business value or user story in every PR. This also auto-closes issues linked to the PR.

**Suggestion:** Use keywords like `fixes`, `closes`, `resolves` and adding the issue number `#123` in PR descriptions to automatically link and close issues when PRs are merged.


To help the devs a PR template could be implemented that checks (among other things) whether a PR is linked to an issue. 

Read more here: [Managing and standardizing pull requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/managing-and-standardizing-pull-requests)

These steps will ensure every feature, bug fix, and hotfix is documented and traceable. 

If we agree I would be happy to contribute such a description to the contributing.md.

**Comment by janhalen** [comment]
@sobuos: If the part where i suggest a PR template seems unclear, click the below simple video explanation of what its all about:


[![Use templates to standardize your Pull Requests](https://img.youtube.com/vi/MjsS4ujX3nU/0.jpg)](https://www.youtube.com/watch?v=MjsS4ujX3nU)

**Comment by andreasDroid** [comment]
Yes, very good idea with the use of keyword. Already started doing this with fix, but will start to use closes/resolves for feature issues. 

Sounds good to me to have a PR template 👍

**Comment by janhalen** [comment]
Ok.. so this issue seems to have branched out a bit 😄 But its all good!  Open discussions of how the product actually should be delivered, are signs of a healthy project.

When looking through the discussion and breaking it down, I find that there is basis for 3 new seperate tasks.

1.  Improve the CONTRIBUTING.md file with workflow and process documentation. 
2.  Implement a PULL_REQUEST_TEMPLATE.MD for the project
3,  Define what automatic checks would benefit the developers and the project, leading to the development and implementation of a pullrequest.yml automation.

If we agree on these tasks, I will add them to the issue-tracker and close this issue.

@sobuos & @andreasDroid, any comments on these suggestions and recommendations.

**Comment by o-droids** [comment]
Closing this issue due to inactivity

**Comment by janhalen** [comment]
@o-droids: I appreciate your work to try and keep the issue-tracker hygiene in check! Thanks for that.. 

One question comes to mind though: Should the closing of this issue as "Completed" be interpreted as: "No plans for creating and getting the added value out of standardized PR_TEMPLATES supporting automation?" or is it just postponed for now?
